project2/kernel.c

```diff
@@ -302,17 +423,21 @@ static void Next_Kernel_Request() {
         Kernel_Request_GetArg,
         Kernel_Request_GetPid,
         Kernel_Request_GetNow,
+        Kernel_Msg_Send,
+        Kernel_Msg_Recv,
+        Kernel_Msg_Rply,
+        Kernel_Msg_ASend,
         Kernel_Request_Terminate
     };
 
     Dispatch();  /* select a new task to run */
 
-    while(1) {
+    for(;;) {
         if (request_info) {
             // Clear it's request
             request_info->request = NONE;
             // Clear our reference to request_info
-            request_info = NULL;
+            /* request_info = NULL; */
```

Comment:
This `NULL` assignment was for cleaning up dangling pointers - at this point we're about to leave the kernel, so we shouldn't need the request_info pointer anymore. It'll get set during entry to the next kernel request

Reply:
Oops, I meant to leave a better comment here. I think the it shouldn't be set to `NULL` because the process descriptor will still have reference to it. For the case a process sends a message and gets put in the blocked state. The kernel will switch to another process for an undetermined amount of time, but when the receiving process finally replies, data needs to be returned to the sending process. So I think fi we set the `request_info` to `NULL` will won't be able to return any data. Same situation when a processes is in blocked state waiting to receive.

Reply:
You're right about all that, but `request_info = NULL` is different from `*request_info = NULL`. The former just 'unattaches' the pointer, the latter sets the data at the location the pointer points at to `NULL`

Reply:
This is so that the `request_info` pointer is not set when the kernel isn't active. This makes it so that if the next kernel request isn't made correctly, we won't get confused and use the pointer from the last request

Reply:
Oh, right 🤦‍♂️  Man C pointer stuff still gets me today... I shall put it back in haha